single_kernel_mongo/state/app_peer_state.py

```diff
@@ -0,0 +1,164 @@
+                if value:
+                    self.data_interface.set_secret(self.relation.id, key, value)
+                else:
+                    self.data_interface.delete_secret(self.relation.id, key)
+            else:
+                self.data_interface.update_relation_data(self.relation.id, {key: value})
+
+    @property
+    def role(self) -> MongoDBRoles:
+        """The role.
+
+        Either from the app databag or from the default from config.
+        """
+        databag_role: str = str(self.relation_data.get(AppPeerDataKeys.role.value))
+        if not self.relation or not databag_role:
+            return self._role
+        return MongoDBRoles(databag_role)
+
+    @role.setter
+    def role(self, value: str) -> None:
+        self.update({"role": value})
+
+    def is_role(self, role_name: str) -> bool:
+        """Checks if the application is running in the provided role."""
+        return self.role == role_name
+
+    @property
+    def db_initialised(self) -> bool:
+        """Whether the db is initialised or not yet."""
+        if not self.relation:
+            return False
+        return json.loads(self.relation_data.get(AppPeerDataKeys.db_initialised.value, "false"))
+
+    @db_initialised.setter
+    def db_initialised(self, value: bool):
+        if isinstance(value, bool):
+            self.update({AppPeerDataKeys.db_initialised.value: json.dumps(value)})
+        else:
+            raise ValueError(
+                f"'db_initialised' must be a boolean value. Provided: {value} is of type {type(value)}"
+            )
+
+    @property
+    def replica_set_hosts(self) -> list[str]:
+        """Returns the stored list of replica set hosts."""
+        if not self.relation:
+            return []
+
+        return json.loads(self.relation_data.get(AppPeerDataKeys.replica_set_hosts.value, "[]"))
+
+    @replica_set_hosts.setter
+    def replica_set_hosts(self, value: list[str]) -> None:
+        self.update({AppPeerDataKeys.replica_set_hosts.value: json.dumps(sorted(value))})
+
+    @property
+    def managed_users(self) -> set[str]:
+        """Returns the stored set of managed-users."""
+        if not self.relation:
+            return set()
+
+        return set(
+            json.loads(self.relation_data.get(AppPeerDataKeys.managed_users_key.value, "[]"))
+        )
+
+    @managed_users.setter
+    def managed_users(self, value: set[str]) -> None:
+        """Stores the managed users set."""
+        self.update({AppPeerDataKeys.managed_users_key.value: json.dumps(sorted(value))})
+
+    @property
+    def keyfile(self) -> str:
+        """Gets the keyfile from the app databag."""
+        if not self.relation:
+            return ""
+
+        return self.relation_data.get(AppPeerDataKeys.keyfile.value, "")
+
+    @keyfile.setter
+    def keyfile(self, keyfile: str):
+        """Stores the keyfile in the app databag."""
+        self.update({AppPeerDataKeys.keyfile.value: keyfile})
```

Comment:
This is a secret (data interfaces make it transparent) but this also disappears in another PR (with config-server/shard) because I didn't like the idea of hiding the accesses to secrets.